### PR TITLE
feat(dex): integrate dex protocol for runtime resources

### PR DESCRIPTION
**Pull request by k-dot-greyz**
## overview

integrates the dex protocol (MIDI 2.0 baseline) into zenOS for runtime resource cataloging.

### dex infrastructure
- **`dex/7E-01_dex-yoinker.py`** - yoinker script for scanning and indexing
- **`dex/README.md`** - integration documentation (0x7E:0x02)
- **`dex/7E-00_dex-index.md`** - auto-generated index (3 entries)

### runtime utilities
- **`zen/utils/dex.py`** - python module for querying dex metadata (0x7C:0x01)
  - `DexReader` class for runtime access
  - methods: `get()`, `by_bank()`, `by_type()`, `search()`
  - lightweight, no external dependencies

## integration with dev-master

follows the dual-layer architecture:
- **dev-master/dex/** - protocol specs and infrastructure docs
- **zenOS/dex/** - runtime resources and operational catalog

both use the same protocol, different domains:
- dev-master: procedures, templates, infrastructure (0x7F, 0x7E, 0x7D)
- zenOS: agents, runtime utilities, workflows (0x7C, 0x7B, runtime)

## protocol reference

see dev-master for full specs:
- [dex protocol spec](https://github.com/k-dot-greyz/dev-master/blob/main/dex/7F-7F_dex-protocol-spec.md)
- [architecture doc](https://github.com/k-dot-greyz/dev-master/blob/main/dex/7E-02_dex-architecture.md)

### scan and index
```bash
python3 dex/7E-01_dex-yoinker.py
cat dex/7E-00_dex-index.md
```

### python api
```python
from zen.utils.dex import DexReader

dex = DexReader()
agents = dex.by_bank(0x7C)  # all agent resources
reader = dex.get('0x7C:0x01')  # specific utility
```

## next steps
- add dex frontmatter to zen/ modules
- integrate into zen cli (`zen dex` commands)
- sync with dev-master dex updates

---

> [!NOTE]
> Adds dex metadata indexing (yoinker script + generated index), a `DexReader` runtime utility, and integration documentation.
> 
> - **Indexing/Ops**:
>   - **Script**: Add `dex/7E-01_dex-yoinker.py` to scan files for dex frontmatter and generate an index.
>   - **Generated Index**: Create `dex/7E-00_dex-index.md` listing entries with `dex_id`, type, status, file link, and `pe_id`.
> - **Runtime Utility**:
>   - **`DexReader`** in `zen/utils/dex.py` to load/query the index (`get()`, `by_bank()`, `by_type()`, `search()`, `list_all()`, `refresh()`), plus `get_dex_metadata()` parser.
> - **Documentation**:
>   - Add `dex/README.md` describing zenOS dex integration, quick start, protocol references, and usage examples.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 4bfe84b55360ca8b009fdb14876440200651eab7. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Adds an automated, sortable catalog index with CLI and programmatic access for listing, searching, filtering, and refreshing entries.

* **Documentation**
  * New guide describing catalog purpose, quick-start commands, usage examples, and indexing scope.

* **Tests**
  * Unit tests covering index parsing, retrieval, filtering, refresh, and metadata extraction edge cases.

* **Bug Fixes**
  * Fixed requirements formatting that could break dependency parsing.

* **Other**
  * Indexing reports progress and skips unreadable or non-catalog files.

### CODE REVIEW

File: dex/7E-01_dex-yoinker.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+# Regex for YAML frontmatter
+YAML_FRONT_RE = re.compile(r"^---\s*\n(.*?)\n---\s*\n", re.DOTALL | re.MULTILINE)
+
+# Keys to extract
+KEYS_TO_YOINK = {
+    "dex_id": re.compile(r"^dex_id:\s*[\"']?(0x[0-9A-F]{2}:0x[0-9A-F]{2})[\"']?", re.MULTILINE),
+    "dex_type": re.compile(r"^dex_type:\s*[\"']?(\w+)[\"']?", re.MULTILINE),
+    "status": re.compile(r"^status:\s*[\"']?(\w+)[\"']?", re.MULTILINE),
+    "pe_id": re.compile(r"property_exchange_id:\s*[\"']?([a-zA-Z0-9:\-\.]+)[\"']?", re.MULTILINE)
+}
```
[the remark this answers is not in the document]
Comment (k-dot-greyz) [reply dex/7E-01_dex-yoinker.py]: @cursor agent In dex/7E-01_dex-yoinker.py around lines 26 to 35, the dex_id regex only matches
uppercase hex digits so lowercase IDs like 0x7e:0x01 are missed; update the
pattern to accept both cases by either adding a-f to the character class
(0-9A-Fa-f) or apply case-insensitive matching (re.IGNORECASE) when compiling
that regex, and ensure any test uses the same flag or pattern style so lowercase
hex IDs are correctly captured.

**Comment by k-dot-greyz** [comment]
> **Actionable comments posted: 0**
> 
> ♻️ Duplicate comments (2)
> > zen/utils/dex.py (2)> `76-101`: **Consider indexing entries by dex_id for O(1) lookup.**
> > > `get()` currently linearly scans `self.entries`, which is fine for the small indexes you have now but may get slower as the dex grows. Building a simple dict mapping `dex_id` → entry during `_load_index()` (and updating it in `refresh()`) would make `get()` constant‑time.
> > > ```diff
> > >  class DexReader:
> > > @@
> > > -        self.entries = []
> > > -        self._load_index()
> > > +        self.entries = []
> > > +        self._index_by_id: Dict[str, Dict] = {}
> > > +        self._load_index()
> > > @@
> > >      def _load_index(self):
> > > @@
> > > -                    self.entries.append(entry)
> > > +                    self.entries.append(entry)
> > > +                    self._index_by_id[dex_id] = entry
> > > @@
> > >      def get(self, dex_id: str) -> Optional[Dict]:
> > >          """get entry by dex_id"""
> > > -        for entry in self.entries:
> > > -            if entry["dex_id"] == dex_id:
> > > -                return entry
> > > -        return None
> > > +        return self._index_by_id.get(dex_id)
> > > @@
> > >      def refresh(self):
> > >          """reload the index"""
> > > -        self.entries = []
> > > -        self._load_index()
> > > +        self.entries = []
> > > +        self._index_by_id = {}
> > > +        self._load_index()
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 
> > > `27-145`: **Add focused tests for DexReader and get_dex_metadata.**
> > > Given this is now the core runtime accessor for dex metadata, unit tests around `_load_index()` (including malformed/empty tables), `get()/by_bank()/by_type()/search()`, and `get_dex_metadata()` (frontmatter vs Python docstring, missing fields, bad encodings) would be valuable. This was already raised in earlier review, but it remains a gap.
> 
> 🧹 Nitpick comments (3)
> > dex/7E-01_dex-yoinker.py (2)> `37-75`: **Frontmatter parsing is robust; consider centralizing regex/patterns.**
> > > The `parse_file()` flow (YAML frontmatter + Python triple‑quoted fallback) looks correct and fails safely. However, the per‑field regexes here mirror those in `zen/utils/dex.get_dex_metadata()`. To avoid drift when the dex schema evolves, consider factoring the shared patterns/frontmatter extraction into a small helper (or shared constants) used by both modules. You could also narrow the `try` block to just the file I/O to satisfy TRY300, leaving the parsing logic outside the `try/except`.
> > > 
> > > `104-120`: **Address Ruff E701 one‑liners for readability (non‑blocking).**
> > > The one‑line `if ...: dirs.remove(...)` and `if ...: continue` statements are flagged by Ruff (E701). Expanding them to multi‑line `if` blocks improves readability and keeps linters quiet, but it’s purely stylistic and doesn’t affect behavior.
> > > ```diff
> > > -        if ".git" in dirs: dirs.remove(".git")
> > > -        if "node_modules" in dirs: dirs.remove("node_modules")
> > > -        if "__pycache__" in dirs: dirs.remove("__pycache__")
> > > +        if ".git" in dirs:
> > > +            dirs.remove(".git")
> > > +        if "node_modules" in dirs:
> > > +            dirs.remove("node_modules")
> > > +        if "__pycache__" in dirs:
> > > +            dirs.remove("__pycache__")
> > > @@
> > > -        for entry in entries:
> > > -        if entry["dex_id"] == "N/A": continue
> > > +        for entry in entries:
> > > +            if entry["dex_id"] == "N/A":
> > > +                continue
> > > ```
> > 
> > 
> > 
> > zen/utils/dex.py (1)> `111-145`: **Deduplicate metadata extraction logic with yoinker.**
> > > `get_dex_metadata()` re‑implements the same frontmatter/triple‑quoted parsing and field regexes as `parse_file()` in `dex/7E-01_dex-yoinker.py`. To keep the dex protocol surface consistent, consider factoring frontmatter detection + the `dex_id`/`dex_type`/`status`/`pe_id` patterns into shared helpers/constants and using them from both places. That way, changes to the dex schema only need to be made once.
> 
> 📜 Review details

@cursoragent how do we approach this? i have copilot working on some of this but i need higher level orchestration here. should I merge this and work on the outstanding issues or should we refine this PR to serve as a solid base for the rest of the work?

**Comment by k-dot-greyz** [comment]
@coderabbitai full review

**Review mode:** please run this like **assertive** (max feedback: potential issues, refactors, security/edge cases, and nitpicks where useful). If assertive is only controllable via `.coderabbit.yaml`, treat this comment as the human override for this PR.

**OSS / free tier — use everything that applies** (per CodeRabbit docs): full walkthrough with changed-file summary, effort estimate, sequence diagrams when they add value, related issues + related PRs, linked-issue fit assessment, suggested labels + reviewer suggestions, review status detail, collapsible walkthrough, and the “prompt for AI agents” style inline hints where supported.

**Finishing touches (if enabled for this repo/plan):** call out missing docstrings / unit-test gaps; if generation commands are allowed, say so explicitly in review comments (I know **autofix** is Pro-only).

Thanks — going for maximum signal without asking for paid-only automation.

**Comment by k-dot-greyz** [comment]
@cursor Please fix failing CI and/or the blocked merge status on this thread so it is green and merge-ready for human approval. Do not merge. Do not close.
